src/lib/HST/Frontend/Transformer/Messages.hs

```diff
@@ -24,3 +26,142 @@ skipNotSupported
 skipNotSupported feature = report
   $ Message Info
   $ feature ++ " are not supported and will be skipped!"
+
+-- | Reports a fatal error that the given feature is not supported. Displays an
+--   excerpt of the input code specified by the given source span.
+notSupportedWithExcerpt
+  :: Members '[InputFile, Report] r
+  => String      -- ^ The name of the feature (plural) that is not supported.
+  -> S.SrcSpan a -- ^ The source span of the excerpt to display.
+  -> Sem r b
+notSupportedWithExcerpt feature srcSpan = do
+  excerpt <- displayCodeExcerpt srcSpan
+  reportFatal $ Message Error $ feature ++ " are not supported!\n" ++ excerpt
+
+-- | Informs the user that the given feature is not supported and the
+--   corresponding AST node will be skipped. Displays an excerpt of the input
+--   code specified by the given source span.
+skipNotSupportedWithExcerpt
+  :: Members '[InputFile, Report] r
+  => String      -- ^ The name of the feature (plural) that is not supported.
+  -> S.SrcSpan a -- ^ The source span of the excerpt to display.
+  -> Sem r ()
+skipNotSupportedWithExcerpt feature srcSpan = do
+  excerpt <- displayCodeExcerpt srcSpan
+  report
+    $ Message Info
+    $ feature ++ " are not supported and will be skipped!\n" ++ excerpt
+
+-- | Displays an excerpt of the input code specified by the given source span.
+--
+--   The excerpt consists of an introductory line with the file path and the
+--   source span numbers, all lines of the input program that are at least
+--   partially contained in the given source span, including their line
+--   numbers, and marks showing the start and end of the spanned code.
+displayCodeExcerpt :: Member InputFile r => S.SrcSpan a -> Sem r String
+displayCodeExcerpt S.NoSrcSpan = return "No source span information available!"
+displayCodeExcerpt src@S.SrcSpan {} = do
+  maybeContent <- getInputFile (S.srcSpanFilePath src)
+  return $ case maybeContent of
+    Nothing      ->
+      "No content was found for the file `" ++ S.srcSpanFilePath src ++ "`!"
+    Just content ->
+      let ls = getLines (S.srcSpanStartLine src - 1) (S.srcSpanEndLine src)
```

Comment:
I think it would be better to put the `-1` into the function `getLines` because the `-1` after `S.srcSpanStartLine src` can be confusing.

Reply:
I wanted `getLines` to be a generic sublist function, so I like it better the way it is now. And I don't think it is that confusing, there are `1`s added to or subtracted from these source span values at many places in this function and its subfunctions.